x-pack/plugins/canvas/public/socket.js

```diff
@@ -8,15 +8,51 @@ import io from 'socket.io-client';
 import { functionsRegistry } from '../common/lib/functions_registry';
 import { getBrowserRegistries } from './lib/browser_registries';
 
+const SOCKET_CONNETION_TIMEOUT = 5000; // timeout in ms
 let socket;
 
-export function createSocket(basePath) {
-  socket = io(undefined, { path: `${basePath}/socket.io` });
+export async function createSocket(basePath) {
+  if (socket != null) return socket;
 
-  socket.on('getFunctionList', () => {
-    const pluginsLoaded = getBrowserRegistries();
+  return new Promise((resolve, rej) => {
+    // status.resolve = resolve;
+    const reject = p => {
+      socket = null; // reset the socket on errors
+      rej(p);
+    };
 
-    pluginsLoaded.then(() => socket.emit('functionList', functionsRegistry.toJS()));
+    socket = io({
+      path: `${basePath}/socket.io`,
+      transports: ['polling', 'websocket'],
+      transportOptions: {
+        polling: {
+          extraHeaders: {
+            'kbn-xsrf': 'professionally-crafted-string-of-text',
+          },
+        },
+      },
+      rememberUpgrade: false,
```

Comment:
You might consider adding a comment here, or above `socket = io({` explaining why this is configured the way it is, perhaps a short link to the API page.  For example, had I not had the conversation with you last night, I wouldn't know why `rememberUpgrade` is set.